[MAOXIONGKAI] iP

MAOXIONGKAI opened the pull request:
> "If you look good, you feel good, If you feel good, you play good, If you play good, they pay good." - Deion Sanders ([source](https://www.on3.com/news/deion-sanders-the-best-quotes-from-prime-time/))

But how do you make yourself look good in the first place? Try Gopher! The best task manager desktop application that you can ever have, because:

- Gopher offers text-based UI that is straightforward and easily readable!
- You can achieve powerful task management just by combine a few commands and keywords together!
- It is more efficient and easier to use!

Too lazy to get started? No worries, all you need to do is just:
1. Download it from [here](https://github.com/MAOXIONGKAI/ip/releases)
2. Run it
3. Add task and boom~
4. We will handle the rest **automatically** for you!

**Our features:**
- [x] Add/Delete task
- [x] Mark/Unmark task
- [x] Find task based on keyword search
- [x] Appealing visual representation of your tasks :star_struck:
- [ ] Query tasks based on time(Coming Soon)
- [ ] Task Reminders(Coming Soon)
- [ ] Appealing GUI Interface(Coming Soon)

Of course, even this application is named after Go programming language's mascot Gopher, we still hope that people like you can write decent Java programs and help contribute back to this project :smiling_face_with_three_hearts:. Here's how you can communicate with me on any coding ideas over here:
```
public static void main(String[] args) {
    System.out.println("Even though I am learning Go, I still like Java");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

MAOXIONGKAI committed:
Change Project Name from Duke to Gopher

MAOXIONGKAI committed:
Change README Project title from Duke to Gopher

MAOXIONGKAI committed:
Gopher.java: Add echo and exit feature

MAOXIONGKAI committed:
Gopher.java: Add addTask feature

MAOXIONGKAI committed:
Gopher.java: Add listTask feature

MAOXIONGKAI committed:
Add Task.java

MAOXIONGKAI committed:
Gopher.java: Add markAsDone and markAsNotDone feature

MAOXIONGKAI committed:
Task.java: change access modifier for name and isDone to protected

MAOXIONGKAI committed:
Add ToDo.java for ToDo Task

MAOXIONGKAI committed:
Add Deadline.java for Task of type Deadline

MAOXIONGKAI committed:
Add Event.java for Task of type Event

MAOXIONGKAI committed:
Gopher.java: Added feature to add ToDo Task

MAOXIONGKAI committed:
Gopher.java: Add feature to add Deadline Task

MAOXIONGKAI committed:
Gopher.java: Add feature to add Event Task

MAOXIONGKAI committed:
Simplify toString implementation for ToDo, Deadline & Event

MAOXIONGKAI committed:
Gopher.java: Add plural check for add task message

MAOXIONGKAI committed:
Set up Test UI Test

MAOXIONGKAI committed:
Add UnknownCommandException.java to handle error when user enters invalid command

MAOXIONGKAI committed:
Add EmptyTaskDescriptionException.java to handle error when user leave out the task description when creating new task

MAOXIONGKAI committed:
Add MissingTokenException.java to handle error when user forgot to enter the keyword needed to create a specific task such as todo, deadline and event

MAOXIONGKAI committed:
Gopher.java: Add try-catch block to handle the newly created exceptions

MAOXIONGKAI committed:
Gohper.java: Add delete task feature

MAOXIONGKAI committed:
Gopher.java: Fix the issue of MissingTokenException giving wrong missing token when creating event task

MAOXIONGKAI committed:
Refactor: Implement Task creation logic using factory method

MAOXIONGKAI committed:
Add more detail to the message for unknown command exception

MAOXIONGKAI committed:
text-ui-test: Improve UI test cases to automate check for error handling

MAOXIONGKAI committed:
Edit .gitignore: Ignore /task folder

MAOXIONGKAI committed:
Add TaskManager.java for saving/loading local task file

MAOXIONGKAI committed:
Change Task to abstract class and add getSaveMessage() abstract method

MAOXIONGKAI committed:
Add task auto-save feature

MAOXIONGKAI committed:
Add FileCorruptedException.java

MAOXIONGKAI committed:
Add basic auto-load task feature

MAOXIONGKAI committed:
Fix the issue of tasks only saved when added new task

MAOXIONGKAI committed:
Fix the issue of task status not retained when saving

MAOXIONGKAI committed:
Fix the issue of empty task file not read properly by the program

MAOXIONGKAI committed:
Merge branch 'branch-Level-7'

MAOXIONGKAI committed:
Add Parser.java to handle LocalDateTime parsing

MAOXIONGKAI committed:
Change the date representation from String to LocalDateTime

MAOXIONGKAI committed:
Update UI test cases to test on date inputs

MAOXIONGKAI committed:
Add Exception handling for invalid date input

MAOXIONGKAI committed:
Update UI test to check for invalid date error handling

MAOXIONGKAI committed:
Merge branch 'branch-Level-8'

MAOXIONGKAI committed:
Change task list from ArrayList of Task to TaskList class

MAOXIONGKAI committed:
Separate the UI Logic into UI.java

MAOXIONGKAI committed:
Add parse methods for add, mark, unmark and delete task command

MAOXIONGKAI committed:
Improve OOP Design for the main program

MAOXIONGKAI committed:
Update UI test to test the new UI layout

MAOXIONGKAI committed:
Merge branch 'branch-A-MoreOOP'

MAOXIONGKAI committed:
Separate files into packages

MAOXIONGKAI committed:
Merge branch 'branch-A-Packages'

MAOXIONGKAI committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

MAOXIONGKAI committed:
Fix run task

MAOXIONGKAI committed:
Merge branch 'branch-A-Gradle'

MAOXIONGKAI committed:
Add unit test for Parser.java

MAOXIONGKAI committed:
Add unit test for Task.java

MAOXIONGKAI committed:
Merge branch 'branch-A-JUnit'

MAOXIONGKAI committed:
Add javadoc for Parser.java

MAOXIONGKAI committed:
Add javadoc for TaskManager.java

MAOXIONGKAI committed:
Add javadoc for Task.java

MAOXIONGKAI committed:
Add javadoc for TaskList.java

MAOXIONGKAI committed:
Add javadoc for UI.java

MAOXIONGKAI committed:
Add javadoc to Gopher.java

MAOXIONGKAI committed:
Fix code standards

MAOXIONGKAI committed:
Add find task feature based on keyword

MAOXIONGKAI committed:
Merge branch 'branch-A-JavaDoc'

MAOXIONGKAI committed:
Merge branch 'branch-A-CodingStandard'

MAOXIONGKAI committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/gopher/parser/Parser.java
#	src/main/java/gopher/task/TaskList.java
#	src/main/java/gopher/ui/UI.java

Resolve: Keep branch-Level-9 changes but adjust the layout for the javadoc

MAOXIONGKAI committed:
Merge branch 'master' into branch-A-JavaDoc

MAOXIONGKAI committed:
Add javadoc for methods used to implement find task feature

MAOXIONGKAI committed:
Merge branch 'branch-A-JavaDoc'

GabrielCWT commented on src/main/java/gopher/Gopher.java:
```diff
@@ -0,0 +1,81 @@
+    private static Scanner inputReader;
+
+    /** TaskList object used by Gopher to track user tasks */
+    private static TaskList taskList;
```
Just a suggestion but maybe you could rename this to tasks

GabrielCWT commented on src/main/java/gopher/parser/Parser.java:
```diff
@@ -0,0 +1,145 @@
+ */
+public class Parser {
+    /**
+     * DateTimeFormater for date input
```
Small typo in comment should be `DateTimeFormatter`

MAOXIONGKAI replied:
Thank you Gabriel for your sharp eyes! I guess I accidentally made this typo and the autocomplete mess up everything else. Thanks for your other inputs as well, I will try rectify them in the next update, thank you!

GabrielCWT commented on src/main/java/gopher/parser/Parser.java:
```diff
@@ -0,0 +1,145 @@
+     * @param command unmark task command
+     * @return task number of the task marked as not done
+     */
+    public static int parseUnMarkCommand(String command) {
```
unmark is one word so `parseUnmarkCommand`

GabrielCWT commented on src/main/java/gopher/storage/TaskManager.java:
```diff
@@ -0,0 +1,135 @@
+            try {
+                Task newTask = null;
+                switch (taskType) {
+                case "T":
```
This could be refactored to be an enum

GabrielCWT reviewed (commented):
Overall good practices. One suggestion is maybe you might want to refactor the tasks into their separate files if they get too big. Well done!

MAOXIONGKAI committed:
Add StyleCheck

MAOXIONGKAI committed:
Fix style issues warned by CheckStyle

MAOXIONGKAI committed:
Merge branch 'branch-A-CheckStyle'

MAOXIONGKAI committed:
Upgrade Gopher from CLI to GUI

MAOXIONGKAI committed:
Add basic styling to Gopher

MAOXIONGKAI committed:
Fix the issue of error message not shown by the GUI interface

MAOXIONGKAI committed:
Merge branch 'branch-Level-10'

MAOXIONGKAI committed:
Apply Variable Arguments to allow deleting multiple tasks simultaneously

MAOXIONGKAI committed:
Apply Variable Arguments to allow marking multiple tasks as done

MAOXIONGKAI committed:
Apply Variable Arguments to allow marking multiple tasks as not done

MAOXIONGKAI committed:
Fix minor typos

MAOXIONGKAI committed:
Merge branch 'branch-A-Varargs'

ArtillerySun commented on src/main/java/gopher/parser/Parser.java:
```diff
@@ -0,0 +1,243 @@
+        return taskNumbers;
+    }
+
+    /**
+     * Parses find task command.
+     *
+     * @param command find task command
+     * @return keyword used for searching
+     */
```
Good java doc writing style!

ArtillerySun commented on src/main/java/gopher/parser/Parser.java:
```diff
@@ -0,0 +1,243 @@
+        }
+
+        StringBuilder taskName = new StringBuilder();
+        if (taskType.equalsIgnoreCase("todo")) {
```
what about using an enum class to simplify the if-else branches?

ArtillerySun commented on src/main/java/gopher/task/Event.java:
```diff
@@ -0,0 +1,44 @@
+ * Include task description, start date and end date.
+ */
+public class Event extends Task {
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
```
clear naming style!

ArtillerySun reviewed (commented):
Good job!

MAOXIONGKAI committed:
Change build name from gopher to Gopher

MAOXIONGKAI committed:
Parser.java: Enhance code quality

The parseCreateTaskCommand is too long and hard to understand.
Some constants are named incorrectly using camel cases.

Use extract method refactoring to shorten method and
change the constant names to SCREAMING_SNAKE_CASE

Such approach improves code quality because it provides higher level of
abstractions which makes the code a lot easier to read and maintain.
Also, using SCREAMING_SNAKE_CASE makes it easier to identify constants

Improve on existing Javadoc and add more comments for clarity

MAOXIONGKAI committed:
Refactor: TaskManager and Parser

The current TaskManager contains logic of interpreting saved task string
which should be inside Parser instead

Shift this part of the logic to Parser to improve organization of code

Some minor adjustments:
* Change constant name from camelCase to SCREAMING_SNAKE_CASE
* Rename some methods for better clarity as the existing ones can result
in confusion

MAOXIONGKAI committed:
TaskList.java: Update comments

TaskList no longer directly prints out messages on UI after introducing
GUI in Level-10 increment, but the comments are not updated

These comments are now removed for better clarity of the program

Some comments that briefly explains on the reason of code implementation
are added for clarity as well

MAOXIONGKAI committed:
Parser.java: Change the name and implementation of isInvalidTaskType()

isInvalidTaskType may result in double negation in the future if we
want to use the method to check if a task type is valid, which may
result in less readable code.

Hence, the function name and implementation is negated to make
everything more straightforward and easier to understand

MAOXIONGKAI committed:
Gopher.java: Refactor the getResponse() method

The method body of getResponse() method is too long and consists of many
low level implementation that would make the code very hard to read.

Use extract method refactoring to isolate out the logics to enhance
readability of the code.

Other minor changes:
* Adhere to coding standard and move task creation logic outside of the
else block, and replace it with error handling logic
* Pass the UnknownCommandException to GUI instead because the error is
now thrown by getResponse Method if execution reaches default branch and
it must be handled externally instead.

MAOXIONGKAI committed:
Parser.java: Fix minor typos

Change the task type in MissingTokenException to lower case instead to
make it consistent with other error messages

MAOXIONGKAI committed:
Merge pull request #1 from MAOXIONGKAI/branch-A-CodeQuality

Merge A-CodeQuality Increment into master branch

MAOXIONGKAI committed:
Merge branch 'master' of https://github.com/MAOXIONGKAI/ip

* 'master' of https://github.com/MAOXIONGKAI/ip:
  Parser.java: Fix minor typos
  Gopher.java: Refactor the getResponse() method
  Parser.java: Change the name and implementation of isInvalidTaskType()
  TaskList.java: Update comments
  Refactor: TaskManager and Parser
  Parser.java: Enhance code quality

MAOXIONGKAI committed:
Merge branch 'master' into branch-A-Assertions

MAOXIONGKAI committed:
Add Assertions to the current codebase

Check if the task save file exists before performing task save/load
related functionalities. Also make sure that task added are of valid
data types are are not null

MAOXIONGKAI committed:
Merge pull request #2 from MAOXIONGKAI/branch-A-Assertions

Branch a assertions

MAOXIONGKAI committed:
Merge branch 'master' into branch-A-Streams

MAOXIONGKAI committed:
TaskList.java: Implement Stream

The current implementation of find method is not very straightforward
because it involves a lot of logics like Regex pattern and Matcher which
makes the code looks messy.

Change the iterative implementation to stream as it offers a more
straightforward pipeline in filtering out the desired tasks that matches
the given pattern. Hence improves the readability of the code

MAOXIONGKAI committed:
Merge pull request #3 from MAOXIONGKAI/branch-A-Streams

Branch a streams

MAOXIONGKAI committed:
Create gradle.yml to set up workflow for A-CI increment

MAOXIONGKAI committed:
Merge pull request #4 from MAOXIONGKAI/branch-A-CI

Create gradle.yml to set up workflow for A-CI increment

MAOXIONGKAI committed:
Implement basic update task feature to fulfill C-Update extension

User can now use update [task number] taskName [tokens] field value to
specify the fields within the task that they want to update.

For example: update 4 Hello /by 2024-05-11 changes the 4-th task's name
in the task list to Hello. If the task is Deadline, then change its
due date to May 11 2024, else a warning message will be prompted by
gopher

MAOXIONGKAI committed:
Add error handling to prevent user from updating task with wrong command

Currently user can include tokens that are not recognized by the task
type as part of the task name, which is not a very good way of handling
updating logic because including token as part of the name leads to
more confusion and can make the app hard to use if user is careless

Hence, checks are added to the update logic of each type of task to
raise error message to the user when they accidentally type in invalid
tokens when updating a task. By reinforcing this we can make the whole
task management app cleaner and less error-prone

MAOXIONGKAI committed:
Refactor: shift the parsing of update command logic into Parser.java

The current parsing logic is in individual task type's update method,
which makes the code very messy and hard to organize.

Hence, the parsing logic are shifted to Parser.java to make the code
more readable and adhere to code quality and standards

MAOXIONGKAI committed:
Merge pull request #5 from MAOXIONGKAI/branch-C-Update

Branch c update

MAOXIONGKAI committed:
Apply another style for gopher warning messages to make UI more unique

The warning messages from gopher is of the same style as other messages,
this makes it quite hard for user to identify whether his/her inputs
are correctly accepted by gopher

Therefore, some checks are implemented to render the gopher dialog box
with red gradient background instead to make sure that it stands out
among other messages, hence improving overall user experience.

MAOXIONGKAI committed:
Updated some error messages so that they get styled correctly on GUI

The current warning message styling only applies to message with text
"Please try again" inside, but some of the current error messages from
gopher do not contain this piece of text, hence they are not styled
as expected

Since "Please try again" should be something exclusive to warning
messages only, it should be acceptable to reinforce all error messages
to contain this piece of text. Therefore, those problematic error
messages are fixed by including this piece of text to make sure that
they are styled correctly on the GUI

MAOXIONGKAI committed:
Adjust the size of the profile images for both user and gopher

The current profile images are unnecessarily big, because the users and
the chatbot are fixed characters in the application, this results in a
lot of valuable spaces in the window wasted.

Hence, the profile images are reduced in size to give more space to the
text and other UI elements.

MAOXIONGKAI committed:
Allow resizing of application window

The layout of the program window is messed up when user attempts to
resize the window.

Changes are made by SceneBuilder configurations to ensure that the
window and the UI elements are anchored properly at the expected
position when user resize the window, so as to enhance user experience

MAOXIONGKAI committed:
Merge branch 'branch-A-BetterGui'

MAOXIONGKAI committed:
Update messages to give Gopher a unique personality

The current messages sent by gopher makes him sounds like a robot, which
makes the application quite monotonous, this may reduce user experience
because the interactions are not very entertaining

Hence, some of the messages are rewritten to make sure that Gopher's
messages are more friendly to the user, so as to enhance overall user
experience

MAOXIONGKAI committed:
Merge branch 'branch-A-Personality'

MAOXIONGKAI committed:
Add error handling when input command has missing or invalid task number

Currently, some of the task command such as mark/unmark, delete, update
does not handle situations when user pass in commands with missing or
invalid task number, such as "delete" or "update hello /by 2024-09-17",
and only an empty message will be shown when such error happens. Even
though this is not crashing the application, it is reducing the user
experience when commands are not input properly.

Therefore, MissingTaskNumberException has been added to handle situation
like this, and when the error is caught by gopher, gopher will warn user
about the missing/invalid number in the given command, hence improving
user experience

MAOXIONGKAI committed:
Change how MissingTaskNameException is thrown for update feature

The current implementation of update does not give error warning in
correct logical order(For example, when inputting "update", user is
prompted with don't know how to update instead of don't know which task
to update), this can lead to confusion of user.

Therefore, the logic of throwing the exception is reorganized such that
everything appears in the correct way, improving overall user experience
when using this feature

MAOXIONGKAI committed:
Add error handling when user inputs task number that does not point to
any valid task in the task list

Currently, the system would break if user enters a task number that
does not point to any valid task in the task list, this would render the
system useless if user are perform operations on multiple tasks at the
same time, hence compromising user experience.

Hence, InvalidTaskNumberException is introduced to make sure that user
gets notified correctly on the task number that causes such issues, such
that they can easily change their command to carry out desired actions.

MAOXIONGKAI committed:
Implement save and load API for taskList Object

Currently save and load task are handled by the Task Manager Object,
which requires reference of the ArrayList of tasks inside the TaskList
Object to work correctly. Such implementation is very inflexible as
TaskList object is only accessible within the Gopher class but its
ArrayList of task cannot be accessed due to encapsulation.

Therefore, save and load task API are implemented within TaskList class
as an application of the Tell, Don't Ask Principle to ensure that Gopher
class can get access to save and load feature for its own task list
without having to break the abstraction barrier.

MAOXIONGKAI committed:
Fix the issue of tasks statuses not updated when
InvalidTaskNumberException is thrown.

Currently, task statuses will not get updated correctly because of
premature return of the functions due to InvalidTaskNumberException,
this makes it very hard for user to keep track of the actual status
of the task list, hence rendering the system useless.

Therefore, finally blocks are added to execute the taskList save API
to ensure that task statuses are always saved regardless of whether
an exception is thrown during the program execution.

MAOXIONGKAI committed:
Fix the issue of Gopher congratulating user when no matched task found

Currently, gopher will display a message congratulating user for not
having pending task when there's no task that appear on the search
result. This message is misleading and should not be there when find
method is executed.

Hence, this message has been extract to UI class and would only appear
when list task method is called and not when the find task method is
called. This ensures that the gopher's message to the user is always
accurate

MAOXIONGKAI committed:
Extend InvalidTokenException check to task creation feature

Currently, user can create task with certain token names if these tokens
are not recognized by the respective task creation command. This may
lead to confusion of user when viewing the task list content.

Therefore, the InvalidTokenException check is extended to all types of
task creation command to ensure that user will be warned that they can't
use tokens for other tasks as part of the task name, hence improving the
overall user experience.

MAOXIONGKAI committed:
Extend InvalidTokenException check to task creation feature

Currently, user can create task with certain token names if these tokens
are not recognized by the respective task creation command. This may
lead to confusion of user when viewing the task list content.

Therefore, the InvalidTokenException check is extended to all types of
task creation command to ensure that user will be warned that they can't
use tokens for other tasks as part of the task name, hence improving the
overall user experience.

MAOXIONGKAI committed:
Merge branch 'branch-A-MoreErrorHandling'

MAOXIONGKAI committed:
Fix the issue of task list delete feature removing the wrong task

Currently the delete feature in task list will remove the first n tasks
if user provides n task numbers, which is not correct. It is found that
the retrieving of task from the list is not working properly because the
tasks are accessed using loop control variable i instead of the supplied
task numbers.

Hence, the delete function is changed to ensure that it accesses the
correct tasks based on the given task numbers instead.

MAOXIONGKAI committed:
Extract out the logic of checking Gopher message type to Message.java

Currently, Gopher's response will be styled as red background if the
response is a warning message that contains text "Please try again".
Even though such implementation works in most cases, it will introduce
styling error in the system if user creates a task with "Please try
again" as part of its name, hence affecting user experience. Also, such
way of recognizing message type and styling is not very extensible and
flexible, hence a better idea would be extract out the logic to another
class to explicitly handle this, so as to provide more flexibility when
similar features need to be developed in the future.

As a result, Message class is created to wrap around the text returned
by Gopher so that other information such as message type can be input
and pass around the program for checks. This offers more flexibility as
compared to the initial approach and hence should be included in the
latest update

MAOXIONGKAI committed:
Add logic to handle FileCorruptedException

Currently, Gopher does not give any explicit message to user when the
file is corrupted or missing, this can render the app useless if the
user does not know how to fix the issue himself/herself.

Therefore, checks and warning messages are added to make sure that user
can be aware of the file corruption issue when performing operators that
relies on save files and storage, so as to improve overall user
experience

MAOXIONGKAI committed:
Shortened messages for multiple task operation command

Currently when multiple tasks are manipulated by one single command,
such as mark, unmark and update, the title of the message will be printed
for each task, which makes the whole message looked very messy and hard
to read.

Modify the logic of creating messages for those commands such that the
title only appears for once per operation instead of per task.

MAOXIONGKAI committed:
Add Ui.png for setting up the product website

MAOXIONGKAI committed:
Merge branch 'master' into branch-A-MoreTesting

MAOXIONGKAI committed:
Update Test Cases for Parser class

MAOXIONGKAI committed:
Update test cases for Task class

MAOXIONGKAI committed:
Remove unnecessary import from Parser Test file

MAOXIONGKAI committed:
Merge branch 'branch-A-MoreTesting'

MAOXIONGKAI committed:
Remove unnecessary update

MAOXIONGKAI committed:
Remove unnecessary import statement

MAOXIONGKAI committed:
Merge branch 'master' of https://github.com/MAOXIONGKAI/ip

MAOXIONGKAI committed:
Create UserGuide for new user

MAOXIONGKAI committed:
Update user guide with more detail

MAOXIONGKAI committed:
Merge branch 'branch-A-UserGuide'

MAOXIONGKAI committed:
Merge branch 'master' into branch-A-UserGuide

MAOXIONGKAI committed:
Fix the issue of new line and table not shown in User Guide

MAOXIONGKAI committed:
Merge branch 'branch-A-UserGuide'

MAOXIONGKAI committed:
User Guide: Adjust image size & new line characters

MAOXIONGKAI committed:
Fix and finalize styling and format of user guide

MAOXIONGKAI committed:
Minor style and content fix for User Guide

MAOXIONGKAI committed:
Fix minor style and content issues in user guide

MAOXIONGKAI committed:
Merge branch 'master' of https://github.com/MAOXIONGKAI/ip

MAOXIONGKAI committed:
Fix the issue of valid duration being treated as invalid

Currently, if user updates an event task with new valid start and end
date, but the new start date is after the current end date, Gopher will
complain that the user has given an invalid date duration, which is not
correct. This is because Gopher only checks whether the start date is
before the current end date or not, and not checking whether it is
before the new end date.

Hence, extra logic is added to determine the new start and end date
before the actual comparison to make sure that Gopher will always
analyze the given duration correctly.

MAOXIONGKAI committed:
Fix coding standards in Javadoc comments

MAOXIONGKAI committed:
Add new comments to improve readability of the code

MAOXIONGKAI committed:
Fix the issue of commands not recognized when user type in extra space

MAOXIONGKAI committed:
Fix the issue of list command not shown in unknown command error message
